File: Musixmatch_lyrics_API_App/app/src/main/java/alexejantonov/com/musixmatch_lyrics_api_app/ui/seach_screen/SearchPresenter.java
package alexejantonov.com.musixmatch_lyrics_api_app.ui.seach_screen;

import com.arellomobile.mvp.InjectViewState;
import com.arellomobile.mvp.MvpPresenter;

import java.util.List;

import alexejantonov.com.musixmatch_lyrics_api_app.MyApplication;
import alexejantonov.com.musixmatch_lyrics_api_app.db.DataBase;
import alexejantonov.com.musixmatch_lyrics_api_app.ui.Base.BaseData;
import alexejantonov.com.musixmatch_lyrics_api_app.utils.DataMergeUtil;

@InjectViewState
public class SearchPresenter extends MvpPresenter<SearchFragmentView> {

	private DataBase dataBase = MyApplication.getDataBase();

	void loadData(String query) {
		List<BaseData> baseData = DataMergeUtil.searchListsMerge(dataBase.getAllArtist(), dataBase.getTracks(), dataBase.getQueryData(query));
		getViewState().showData(baseData, query);
	}
}

File: Musixmatch_lyrics_API_App/app/src/main/java/alexejantonov/com/musixmatch_lyrics_api_app/ui/seach_screen/SearchPresenter.kt
package alexejantonov.com.musixmatch_lyrics_api_app.ui.seach_screen

import alexejantonov.com.musixmatch_lyrics_api_app.MyApplication
import alexejantonov.com.musixmatch_lyrics_api_app.api.entities.artist.Artist
import alexejantonov.com.musixmatch_lyrics_api_app.api.entities.track.Track
import alexejantonov.com.musixmatch_lyrics_api_app.ui.Base.BaseData
import alexejantonov.com.musixmatch_lyrics_api_app.utils.DataMergeUtil
import android.util.Log
import com.arellomobile.mvp.InjectViewState
import com.arellomobile.mvp.MvpPresenter
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.functions.BiFunction
import io.reactivex.schedulers.Schedulers

@InjectViewState
class SearchPresenter : MvpPresenter<SearchFragmentView>() {

  private val dataBase = MyApplication.dataBase
  private val subscriptions = CompositeDisposable()

  fun loadData(query: String) {
    subscriptions.add(
        Observable
            .combineLatest(
                Observable.fromCallable { dataBase.allArtist },
                Observable.fromCallable { dataBase.tracks.toMutableList() },
                BiFunction<List<Artist>, MutableList<Track>, List<BaseData>> { artists, tracks ->
                  DataMergeUtil.searchListsMerge(artists, tracks, query)
                }
            )
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe(
                { viewState.showData(it, query) },
                { Log.e("Search failed: ", Log.getStackTraceString(it)) }
            )
    )
  }

  override fun onDestroy() {
    subscriptions.clear()
    super.onDestroy()
  }
}
